src/nvim/ex_cmds2.c

```diff
@@ -2161,48 +2161,8 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
   funccal_entry_T funccalp_entry;
   save_funccal(&funccalp_entry);
 
-  // Check if this script was sourced before to finds its SID.
-  // If it's new, generate a new SID.
-  // Always use a new sequence number.
   const sctx_T save_current_sctx = current_sctx;
-  current_sctx.sc_seq = ++last_current_SID_seq;
-  current_sctx.sc_lnum = 0;
-  FileID file_id;
-  bool file_id_ok = os_fileid((char *)fname_exp, &file_id);
-  assert(script_items.ga_len >= 0);
-  for (current_sctx.sc_sid = script_items.ga_len; current_sctx.sc_sid > 0;
-       current_sctx.sc_sid--) {
-    si = &SCRIPT_ITEM(current_sctx.sc_sid);
-    // Compare dev/ino when possible, it catches symbolic links.
-    // Also compare file names, the inode may change when the file was edited.
-    bool file_id_equal = file_id_ok && si->file_id_valid
-                         && os_fileid_equal(&(si->file_id), &file_id);
-    if (si->sn_name != NULL
-        && (file_id_equal || fnamecmp(si->sn_name, fname_exp) == 0)) {
-      break;
-    }
-  }
-  if (current_sctx.sc_sid == 0) {
-    current_sctx.sc_sid = ++last_current_SID;
-    ga_grow(&script_items, (int)(current_sctx.sc_sid - script_items.ga_len));
-    while (script_items.ga_len < current_sctx.sc_sid) {
-      script_items.ga_len++;
-      SCRIPT_ITEM(script_items.ga_len).sn_name = NULL;
-      SCRIPT_ITEM(script_items.ga_len).sn_prof_on = false;
-    }
-    si = &SCRIPT_ITEM(current_sctx.sc_sid);
-    si->sn_name = fname_exp;
-    fname_exp = vim_strsave(si->sn_name);  // used for autocmd
-    if (file_id_ok) {
-      si->file_id_valid = true;
-      si->file_id = file_id;
-    } else {
-      si->file_id_valid = false;
-    }
-
-    // Allocate the local script variables to use for this script.
-    new_script_vars(current_sctx.sc_sid);
-  }
+  si = get_current_script_id(fname_exp, &current_sctx);
```

Comment:
i think it should be fine to just do something like:

```c
si = get_current_script_id(fname_exp, &current_sctx);
fname_exp = vim_strsave(si->sn_name);  // used for autocmd
```

and change `get_current_script_id()` so that it doesn't allocate.

Reply:
Done.

Reply:
We get a memory leak with this because we are making copy always but we don't need to make a copy if the script already has an sid. A  solution would be to free a copy when script already has an id . This seems hacky .

Are you confident previous way has a good enough chance of causing bug ?

Also since `fname_exp` gets cleared at end of `do_source` Won't global `sourceing_name` hold a invalid pointer after that in this case ?